okhttp/src/main/java/com/squareup/okhttp/internal/http/RouteSelector.java

```diff
@@ -66,27 +61,22 @@
   private List<InetSocketAddress> inetSocketAddresses = Collections.emptyList();
   private int nextInetSocketAddressIndex;
 
-  /* Specs to attempt with the connection. */
-  private List<ConnectionSpec> connectionSpecs = Collections.emptyList();
-  private int nextSpecIndex;
-
   /* State for negotiating failed routes */
   private final List<Route> postponedRoutes = new ArrayList<>();
 
-  private RouteSelector(Address address, URI uri, OkHttpClient client, Request request) {
+  private RouteSelector(Address address, URI uri, OkHttpClient client) {
     this.address = address;
     this.uri = uri;
     this.client = client;
     this.routeDatabase = Internal.instance.routeDatabase(client);
     this.network = Internal.instance.network(client);
-    this.request = request;
 
     resetNextProxy(uri, address.getProxy());
   }
 
   public static RouteSelector get(Address address, Request request, OkHttpClient client)
       throws IOException {
```

Comment:
It turns out that we should just get rid of this IOException completely, which gets rid of a lot of the motivation for RequestException.

This only comes up when the URL is malformed when converted to a URI. We could either detect that eagerly, or possibly improve `toUriLenient` to not throw.

I'd love to not need RequestException if we can avoid it. Our code is better spent on more important problems.